guide/src/free-threading.md

```diff
@@ -0,0 +1,153 @@
+safety around mutably aliased rust variables the Python interpreter can
+access. We believe that it would require adding an `unsafe impl` for `Send` or
+`Sync` to trigger this behavior in code using PyO3. Please report any issues
+related to runtime borrow checker errors on mutable pyclass implementations that
+do not make strong assumptions about the GIL.
+
+It was always possible to generate panics like this in PyO3 in code that
+releases the GIL with `allow_threads` (see [the docs on interior
+mutability](./class.md#bound-and-interior-mutability),) but now in free-threaded
+Python there are more opportunities to trigger these panics because there is no
+GIL.
+
+We plan to allow user-selectable semantics for mutable pyclass definitions in
+PyO3 0.24, allowing some form of opt-in locking to emulate the GIL if that is
+needed.
+
+## `GILProtected` is not exposed
+
+`GILProtected` is a PyO3 type that allows mutable access to static data by
+leveraging the GIL to lock concurrent access from other threads. In
+free-threaded Python there is no GIL, so you will need to replace this type with
+some other form of locking. In many cases, a type from `std::sync::Atomic` or
+a `std::sync::Mutex` will be sufficient.
+
+Before:
+
+```rust
+# fn main() {
+# #[cfg(not(Py_GIL_DISABLED))] {
+# use pyo3::prelude::*;
+use pyo3::sync::GILProtected;
+use pyo3::types::{PyDict, PyNone};
+use std::cell::RefCell;
+
+static OBJECTS: GILProtected<RefCell<Vec<Py<PyDict>>>> =
+    GILProtected::new(RefCell::new(Vec::new()));
+
+Python::with_gil(|py| {
+    // stand-in for something that executes arbitrary Python code
+    let d = PyDict::new(py);
+    d.set_item(PyNone::get(py), PyNone::get(py)).unwrap();
+    OBJECTS.get(py).borrow_mut().push(d.unbind());
+});
+# }}
+```
+
+After:
+
+```rust
+# use pyo3::prelude::*;
+# fn main() {
+use pyo3::types::{PyDict, PyNone};
+use std::sync::Mutex;
+
+static OBJECTS: Mutex<Vec<Py<PyDict>>> = Mutex::new(Vec::new());
+
+Python::with_gil(|py| {
+    // stand-in for something that executes arbitrary Python code
+    let d = PyDict::new(py);
+    d.set_item(PyNone::get(py), PyNone::get(py)).unwrap();
+    // we're not executing Python code while holding the lock, so GILProtected
+    // was never needed
+    OBJECTS.lock().unwrap().push(d.unbind());
+});
+# }
+```
+
+If you are executing arbitrary Python code while holding the lock, then you will
+need to use conditional compilation to use `GILProtected` on GIL-enabled Python
+builds and mutexes otherwise. If your use of `GILProtected` does not guard the
+execution of arbitrary Python code or use of the CPython C API, then conditional
+compilation is likely unnecessary since `GILProtected` was not needed in the
+first place and instead rust mutexes or atomics should be preferred. Python 3.13
+introduces `PyMutex`, which releases the GIL while the lock is held, so that is
+another option if you only need to support newer Python versions.
```

Comment:
Is this interaction between `PyMutex` and the GIL true? I thought it just released the GIL while blocking.

```suggestion
first place and instead rust mutexes or atomics should be preferred. Python 3.13
introduces `PyMutex`, which releases the GIL while the waiting for the lock, so that is
another option if you only need to support newer Python versions.
```

One option is that we add `MutexExt` traits to `pyo3::sync` which work for all Python versions, similar to #4513 

e.g. `.lock_py_attached(py)`

🤔

Reply:
Thanks for spotting the error, I applied your correction.

The extension trait is an interesting idea, it's certainly possible for PyO3 to detach from the GIL if a mutex is blocked. Definitely something to consider...